src/lib/pipe_lib/broadcast_pipe.ml

```diff
@@ -0,0 +1,134 @@
+open Core_kernel
+open Async_kernel
+
+type 'a t =
+  { mvar: 'a option Mvar.Read_only.t
+  ; mutable cache: 'a option
+  ; mvar_pipe_handle: 'a option Pipe.Reader.t
+  ; mutable id: int
+  ; pipes:
+      ( 'a option Strict_pipe.Reader.t
+      * ( 'a option
+        , Strict_pipe.drop_head Strict_pipe.buffered
+        , unit )
+        Strict_pipe.Writer.t )
+      Int.Table.t }
+
+let create mvar =
+  let mvar_pipe_handle = Mvar.pipe_when_ready mvar in
+  let t =
+    { mvar
+    ; cache= Mvar.peek mvar |> Option.join
+    ; mvar_pipe_handle
+    ; id= 0
```

Comment:
variable name is confusing. this isn't the id of the broadcast pipe, it's a counter for the ids of the readers

Reply:
Agreed. Something like `max_id`, `id_counter`, or something would be better.